### ISSUE TITLE
Automatically normalize parameters in MaximumLikelihoodHistogramFitter

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#163]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/163) by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

I think we've all encountered situations where a scipy linear solve in `MaximumLikelihoodHistogramFitter` reports a poorly conditioned matrix. I found that - at least for my fitting tasks - this can be eliminated by normalizing all fit parameters so that they are roughly the same order of magnitude.

What I did was to use `MaximumLikelihoodHistogramFitter` as follows. 
```
#!python

def wrap_func(func, pnorm):
    def wrapped_func(params, x):
        return func(params*pnorm, x)
    return wrapped_func

def model_func(params, x):
    # All sorts of code

parms0 = np.array([<initial parameter>])
prorm = np.array([<values to normalize by>])

fitter = MaximumLikelihoodHistogramFitter(x, n, parms0/pnorm, wrap_func(func, pnorm))
(parms_normed, pcov_normed) = fitter.fit()
parm = parms_normed*pnorm
pcov = np.matmul(np.diag(pnorm), np.matmul(pcov_normed, np.diag(pnorm)))
```
We could easily add a new parameter, `pnorm` to `MaximumLikelihoodHistogramFitter` to make the details transparent to the user. In practice we would nearly always set `pnorm=parms0` so maybe if you set `pnorm=True` that means to use the initial parameter values?

Joe is rightly concerned that this doesn't make so much sense for a value that is centered around zero (or maybe even for one that could be both positive and negative), but I argue that we make the default behavior no normalization,  that's not a concern.

Another concern is what to do when the user passes in a `theory_gradient`. It's not clear to me whether you can just wrap the `theory_gradient` function the same way and expect it to work. I will try to figure this out.

**Comment by joefowler** [comment]
**Original comment by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

Added code to calculate proper covariance matrix

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

It seems to me that we can let the generic `MaximumLikelihoodHistogramFitter` have arguments that control this behavior. But for the line-specific fitters (e.g., `MnKAlphaFitter`) I expect we can specify the rescaling for all instances to be the same.

This example raises an important question. Think about the energy center--the peak location--in Mn Kα fitting problem. It uses peak-energy parameters whose value will be around 5900, but the uncertainty on them can be 5 orders of magnitude smaller. (In many fits, the units are pulse heights instead of energy, but the 5 orders of magnitude still applies.) What is needed to get the problematic matrix well-conditioned? Parameters that are all of order 1, as you've suggested, or parameters whose *uncertainty* is all of the same order?

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

@{557058:34fe21d7-7863-4cd7-88d7-7f2a5c8eb7c2} : So@{557058:9524f6a0-ef9f-4466-b913-d04e51ff9093} and I have decided that this will probably be fixed, implicitly, by the switch to using LMFIT for our fits. This is nearly ready \(watch branch LMFit or issue #165 for the latest if you’re dying to check it out now\).

If we’re wrong about this guess, then you \(or someone\) can just file a new issue. But this is exactly the sort of thing that argues for switching to LMFIT: so that we can take advantage of code crafted by careful experts in the fitting domain, who have spent many person-years perfecting the underlying `scipy.optimize` library.

**Comment by joefowler** [comment]
**Original comment by Dan Becker (Bitbucket: [danbek](https://bitbucket.org/danbek), GitHub: [danbek](https://github.com/danbek)).**

----------------------------------------

sounds good to me

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

This is fixed by PR #113, use of LMFIT. At least, we think it is.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Removing milestone: v0.7 (automated comment)
